readme: add list of licenses - not Apache 2.0

0xc0170 opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->

Add paragraph to readme with all folders containing non Apache 2.0. I would like all teams to verify this and approve.

I am still identifying targets licenses, we might extend that list to be more accurate.

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!-- 
    Optional
    Request additional reviewers with @username
-->

@SeppoTakalo @AnttiKauppila @donatieng @dannybenor @SenRamakri

0xc0170 committed:
readme: add list of licenses - not Apache 2.0

Name all folders containing different liceses. Any new file not Apache 2.0 should
be captured here.

0xc0170 commented:
As soon as this is approved, will add docs team to review and make sure this is also mentioned in our docs (license page). This list will need to be kept up to date.

0xc0170 commented:
> @SeppoTakalo @AnttiKauppila @donatieng @dannybenor @SenRamakri

Added you as reviewers based on the code areas - please look at them, if we have it correctly specified.

SeppoTakalo commented on README.md:
```diff
@@ -33,6 +33,28 @@ The software is provided under [Apache-2.0 license](LICENSE). Contributions to t
+- features/FEATURE_BLE/targets - BSD-style/PBL/MIT
+- features/lorawan - Revised BSD
+- features/lwipstack - BSD-3-Clause/MIT
+- features/nanostack/sal-stack-nanostack - BSD-3-Clause
```
This is only partially true..
Most of the Nanostack itself is Apache 2.0, but there are Thread related code which is BSD.
https://github.com/ARMmbed/mbed-os/tree/master/features/nanostack/sal-stack-nanostack/source/6LoWPAN/Thread

JanneKiiskila replied:
I would recommend doing a proper table with links to the sources/licence file in the three. Makes it easier to find those, since for the BSD related stuff - the end product needs to show the license SOMEWHERE in the UI/manuals.

Having a link to POINT to a direct LICENSE -file also ensures we do our due diligence and include the license in the source code itself, like we need to.

0xc0170 replied:
I stated this lists contains folders with non-apache files (thus nanostack folder marked as it contains BSD-3-Clause, does not mean there are no files with Apache 2.0). Each of these folders should have then list of non Apache files in its folder.

I like the idea having the tree link

> features/nanostack/sal-stack-nanostack - BSD-3-Clause

This would have url link so we can go to nanostack folder and there would be: 
- license file (in this case BSD-3-Clause)
- Readme with license information (what files are non-apache)

In cases like nanostack this is simple, when we consider targets like we have in BLE or under targets/ the list will be very large (multiple licenses involved).

Open for suggestions

0xc0170 replied:
Updated.  The latest push contains links to folders. Each of them should contain Readme (will be added later by separate PRs). I'll add this requirement to our documentation first and will request each module to add it later.

Readme.md in the root should just point to `features/nanostack` (license overview). 
`features/nanostack/Readme.md` will contain list of files/modules inside with licenses (detailed view).

0xc0170 replied:
Our license page actually contains this, we just need to fix the current ones and check the new ones as a requirement

https://os.mbed.com/docs/mbed-os/v5.11/contributing/license.html (`Using a different license`).

SeppoTakalo commented on README.md:
```diff
@@ -33,6 +33,28 @@ The software is provided under [Apache-2.0 license](LICENSE). Contributions to t
+- features/storage/blockdevice - MIT
+- features/storage/filesystem/littlefs - BSD-3-Clause
+- features/storage/filesystem/fat - MIT/BSD-style
+- features/netsocket/emac-drivers - BSD-3-Clause
```
This depends on the driver itself.
There drivers are supplied mostly by our partners, so it is not our choice what license they use.

0xc0170 replied:
We just need to document the licenses - provide a list of files that are not. This is going that direction and we start documenting every change to this (if not Apache files being added, they need to be documented and questioned why not apache 2.0/pbl as our license guide suggests).

SeppoTakalo reviewed (changes requested):
I think the idea of having this list if good, it clarifies things, but the list itself is inaccurate.

Provided some feedback of what should be changed.

ciarmcom commented:
@0xc0170, thank you for your changes.
@AnttiKauppila @SeppoTakalo @SenRamakri @dannybenor @donatieng @ARMmbed/mbed-os-maintainers please review.

cmonr commented:
@0xc0170 Any updates?

AnttiKauppila reviewed (approved):
WAN team content is fine

0xc0170 commented:
@cmonr Yes, I've completed the scan and its results, will update this.

0xc0170 committed:
readme: license update

Fix licenses based on latest Fossology scan. Using -style licenses for specific type licenses
Add paths to the readme.

0xc0170 committed:
readme: storage license under one folder

0xc0170 commented:
I've updated the Readme - used 5.11.3 release scan results. Added paths to each folder as @JanneKiiskila suggested. It now contains components. Each of these will need own readme to specify licenses in detail. This will be the second step. Good exercise will be to capture targets folders.

Ready for review

0xc0170 commented:
Created https://github.com/ARMmbed/mbed-os/issues/9683 to have licenses fixed in each of the items in the list here. Will be tracked separately.

cmonr commented on README.md:
```diff
@@ -33,6 +33,26 @@ The software is provided under [Apache-2.0 license](LICENSE). Contributions to t
 
+Folders containing files not under Apache 2.0 as listed above (each folder should contain own README file with license specified for its files). The original license text is included in those source files.
+
+- [cmsis](./cmsis) - MIT, BSD-3-Clause
```
I _really_ don't like how much this changes what is suppose to be a simple README.

cmonr reviewed (changes requested):
Find a different way to indicate/list/link files with different licenses.

README.md should be short and easy to digest. This PR asis would double the size of the file while only providing very specific information that most users would not care about.

cmonr commented:
Added @ARMmbed/mbed-docs for review (bad bot, leaving them out...)

0xc0170 commented:
> README.md should be short and easy to digest. This PR asis would double the size of the file while only providing very specific information that most users would not care about.

I can create LICENSE file with this list, and just link it here.

0xc0170 committed:
license: rename to apache 2.0 txt file

0xc0170 committed:
license: add license file containing folders with non apache license

0xc0170 commented:
Pushed an update, following our license guide. LICENSE provides overview of licenses, LICENSE-apache-2.0.txt the main license text. And our readme just references both.

0xc0170 committed:
readme: license file reference fix

cmonr commented on README.md:
```diff
@@ -29,10 +29,12 @@ The [release notes](https://os.mbed.com/releases) detail the current release. Yo
-The software is provided under [Apache-2.0 license](LICENSE). Contributions to this project are accepted under the same license. Please see [contributing.md](CONTRIBUTING.md) for more info. 
+The software is provided under [Apache-2.0 license](LICENSE-apache-2.0.txt). Contributions to this project are accepted under the same license. Please see [contributing.md](CONTRIBUTING.md) for more info. 
 
 This project contains code from other projects. The original license text is included in those source files. They must comply with our [license guide](https://os.mbed.com/docs/latest/reference/license.html)
```
```suggestion
This project contains code from other projects. The original license text is included in those source files. They must comply with our [license guide](https://os.mbed.com/docs/latest/reference/license.html).
```

0xc0170 replied:
Fixed

cmonr reviewed (approved):
👍

0xc0170 committed:
readme: fix space and . in license

0xc0170 commented:
Needs more reviews ! I would like to have this on master, so any new non apache code is documented.

0xc0170 commented on LICENSE:
```diff
@@ -1,165 +1,22 @@
+Unless specifically indicated otherwise in a file, files are licensed under the Apache 2.0 license,
+as can be found in: LICENSE-apache-2.0.txt
+
+Folders containing files under different permissive license than Apache 2.0 are listed below. Eeach folder should contain own README file with license specified for its files. The original license text is included in those source files.
```
Eeach -> typo, @AnotherButler Can you fix ? I've just noticed this,  I expect additional styling commit :)

Commit:
Fix typo in LICENSE

Fix typo as requested in comments.

Commit:
Add articles to README.md

Include articles for clarity.
